p2p/client.go

```diff
@@ -121,6 +126,20 @@ func (c *Client) startWithHost(ctx context.Context, h host.Host) error {
 		c.logger.Info("listening on", "address", fmt.Sprintf("%s/p2p/%s", a, c.host.ID()))
 	}
 
+	if c.gater != nil {
```

Comment:
double time checking of `c.gater != nil` really feels unnecessary. can you skip the check here and let the function check for `nil` and return the error?